frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/repository/RepositoryGraphBuilder.java

```diff
@@ -36,147 +39,135 @@
+      if (!isFirstBranch) {
+        graphNodes.get(previousBranchIndex).setNextSiblingNodeIndex(graphNodes.size());
+      }
 
-      int splittingElementIndex = graphElements.size() - 1;
-      buildCommitsAndNonRootBranch(graphElements, branch, /* upElementIndex */ splittingElementIndex,
-          syncToParentStatus);
+      int prevSiblingNodeIndex = graphNodes.size() - 1;
+      assert prevSiblingNodeIndex >= 0;
+      buildCommitsAndNonRootBranch(graphNodes, branch, prevSiblingNodeIndex, indentLevel);
 
-      upElementIndex = graphElements.size() - 2;
-      int upstreamBranchIndex = graphElements.size() - 1;
+      int upBranchIndex = graphNodes.size() - 1;
       List<BaseGitMacheteNonRootBranch> branches = branch.getDownstreamBranches();
-      recursivelyAddCommitsAndBranches(graphElements, /* downstream */ branches, upstreamBranchIndex);
+      recursivelyAddCommitsAndBranches(graphNodes, positionsOfVisibleEdges, /* downstream */ branches,
+          upBranchIndex, indentLevel + 1);
+
+      while (positionsOfVisibleEdges.size() < graphNodes.size()) {
+        positionsOfVisibleEdges.add(new SmartList<>());
+      }
+      if (!branch.equals(lastDownstreamBranch)) {
+        for (int i = upBranchIndex + 1; i < graphNodes.size(); ++i) {
+          positionsOfVisibleEdges.get(i).add(indentLevel);
+        }
+      }
+
+      previousBranchIndex = upBranchIndex;
+      isFirstBranch = false;
     }
-
-    addPhantomGraphElementIfNeeded(graphElements, branchIndex, upElementIndex);
   }
 
-  private void addRootBranch(
-      java.util.List<IGraphElement> graphElements,
-      BaseGitMacheteRootBranch branch) {
-    BranchElement element = createBranchElementFor(branch, /* upElementIndex */ -1, GraphEdgeColor.GREEN,
-        branch.getSyncToRemoteStatus());
-    graphElements.add(element);
+  private void addRootBranch(java.util.List<IGraphNode> graphNodes, BaseGitMacheteRootBranch branch) {
+    BranchNode branchNode = createBranchNodeFor(branch, /* prevSiblingNodeIndex */ -1,
+        GraphEdgeColor.GREEN, /* indentLevel */ 0);
+    graphNodes.add(branchNode);
   }
 
   private void buildCommitsAndNonRootBranch(
-      java.util.List<IGraphElement> graphElements,
+      java.util.List<IGraphNode> graphNodes,
       BaseGitMacheteNonRootBranch branch,
-      int upstreamBranchIndex,
-      SyncToParentStatus syncToParentStatus) {
+      @NonNegative int upstreamBranchIndex,
+      @NonNegative int indentLevel) {
     List<IGitMacheteCommit> commits = branchGetCommitsStrategy.getCommitsOf(branch).reverse();
 
+    var syncToParentStatus = branch.getSyncToParentStatus();
     GraphEdgeColor graphEdgeColor = SyncToParentStatusToGraphEdgeColorMapper.getGraphEdgeColor(syncToParentStatus);
-    ISyncToRemoteStatus syncToRemoteStatus = branch.getSyncToRemoteStatus();
-    int branchElementIndex = graphElements.size() + commits.size();
+    int branchNodeIndex = graphNodes.size() + commits.size();
+    assert branchNodeIndex > 0;
```

Comment:
Drop a comment why `graphNodes` can't be empty here (we unfortunately can't enforce that statically.... mutable collections :cry:)